07_web_endpoints/discord_bot.py

```diff
@@ -0,0 +1,230 @@
+# ---
+# lambda-test: false
+# ---
+
+# # Serve a Discord Bot on Modal
+
+# (quick links: [try it out on Discord](https://discord.gg/nR96BxPu))
+
+# In this example we will demonstrate how to build a Discord bot that uses
+# [slash commands](https://discord.com/developers/docs/interactions/application-commands)
+# and serve it from Modal.
+
+# Slash commands send information from Discord server members to a service at a URL.
+# Here, we set up a simple [FastAPI app](https://fastapi.tiangolo.com/)
+# to run that service and deploy it easily and serverlessly using Modal’s
+# [`@asgi_app`](https://modal.com/docs/guide/webhooks#serving-asgi-and-wsgi-apps) decorator.
+
+# As our example application, we build a simple weather service.
+
+# ## Create a Discord app
+
+# To connect our model to a Discord bot, we’re first going to create an
+# application on the Discord Developer Portal.
+
+# 1. Go to the
+#    [Discord Developer Portal](https://discord.com/developers/applications) and
+#    login with your Discord account.
+# 2. On the portal, go to **Applications** and create a new application by
+#    clicking **New Application** in the top right next to your profile picture.
+# 3. [Create a custom Modal secret](https://modal.com/docs/guide/secrets) for your Discord bot.
+#    On Modal's secret creation page, select 'Discord'. Copy your Discord application’s
+#    **Public Key** (in **General Information**) and paste the value of the public key
+#    as the value of the `DISCORD_PUBLIC_KEY` environment variable.
+#    Name this secret `weather-discord-secret`.
+
+# ## Register a Slash Command
+
+# Next, we’re going to register a command for our Discord app via an HTTP
+# endpoint.
+
+# Run the following command in your terminal, replacing the appropriate variable
+# inputs. `BOT_TOKEN` can be found by resetting your token in the application’s
+# **Bot** section, and `CLIENT_ID` is the **Application ID** available in
+# **General Information**.
+
+
+# ``` shell
+# BOT_TOKEN='replace_with_bot_token'
+# CLIENT_ID='replace_with_client_token'
+# curl -X POST \
+# -H 'Content-Type: application/json' \
+# -H "Authorization: Bot $BOT_TOKEN" \
+# -d '{
+#   "name":"get_weather",
+#   "description":"get weather",
+#   "options":[
+#     {
+#       "name":"city",
+#       "description":"The city for which you want to get the weather",
+#       "type":3,
+#       "required":true
+#     }
+#   ]
+# }' "https://discord.com/api/v10/applications/$CLIENT_ID/commands"
+# ```
+
+
+# This will register a Slash Command for your bot named `get_weather`, and has a
+# parameter called `city`. More information about the
+# command structure can be found in the Discord docs
+# [here](https://discord.com/developers/docs/interactions/application-commands).
+
+# ## Defining the asgi app with modal
```

Comment:
There's a lot of code in between this title/description and the actual ASGI app creation.

Even apart from splitting up the logic further, the prose and code organization also needs to be split up further. The logic for what our service does is separate from how we serve it -- even though our service here is pretty simple.